### PR TITLE
[ML] Reduce false positives associated with the multi-bucket feature

### PR DESCRIPTION
**Pull request by tveasey**
This improves the interaction between multi-bucket feature and the model we maintain attending specifically to periods flagged as anomalous. In particular, we were suppressing updates of this anomaly model if the bucket probability was high but continuing to adjust the overall probability if the multi-bucket feature probability was low. The upshot our scores were inappropriately high.

This change pegs the time used to compute the anomaly duration to the last update time for the anomaly model and also smoothly reduces the impact on our overall result if the bucket value itself is normal.

This should help with #477.

### CODE REVIEW

File: lib/maths/CTimeSeriesModel.cc [hunk 1]
```diff
@@ -387,10 +399,15 @@ class CTimeSeriesAnomalyModel {
...
     //! Update the anomaly model with a sample of the current feature vector.
-    void sample(const CModelProbabilityParams& params, core_t::TTime time, double weight);
+    void sample(const CModelProbabilityParams& params, double weight);
 
     //! Compute the probability of the anomaly feature vector.
-    bool anomalyProbability(core_t::TTime time, double& result) const;
+    bool anomalyProbability(double& result) const;
+
+    //! Get the largest probability the model counts as anomalous.
+    double largestAnomalyProbability() const {
```
Comment (edsavage) [inline lib/maths/CTimeSeriesModel.cc]: A bit of a nit pick - feel free to disregard! - Rename this method to say `twiceLargestSignificantProbability`?

File: lib/maths/CTimeSeriesModel.cc [hunk 2]
```diff
@@ -573,11 +598,13 @@ bool CTimeSeriesAnomalyModel::acceptRestoreTraverser(const SModelRestoreParams&
...
+    // else we can't upgrade the state of the anomaly model pre 6.5.
+
     return true;
 }
 
 void CTimeSeriesAnomalyModel::acceptPersistInserter(core::CStatePersistInserter& inserter) const {
-    inserter.insertValue(VERSION_6_5_TAG, "");
+    inserter.insertValue(VERSION_7_2_TAG, "");
     if (m_Anomaly) {
         inserter.insertLevel(ANOMALY_6_5_TAG, boost::bind(&CAnomaly::acceptPersistInserter,
```
Comment (droberts195) [inline lib/maths/CTimeSeriesModel.cc]: This is no longer restored, so is it worth persisting it?  Or is it a mistake not to restore it?

**Comment by droberts195** [comment]
Also, I'm not convinced it's a good idea to put this into 7.2.1.  This change affects results and 7.2.1 is going to be a patch release that people will upgrade to expecting no change to functionality.  Potentially Cloud clusters could be force-upgraded to 7.2.1 if there turns out to be a security bug in 7.2.0.  Technically this is a bug fix, but if there's some set of circumstances where the anomaly detection is clearly worse after this change then we'll still have a very annoyed user.

Therefore I would say that either we prioritise testing this next Monday and Tuesday and get it into 7.2.0 or else it waits for 7.3.0.

And if it waits for 7.3.0 then the 7.2 constant in the code should be changed.

File: lib/maths/CTimeSeriesModel.cc (hunk 1 above)
Comment (tveasey) [reply lib/maths/CTimeSeriesModel.cc]: I think I'll leave this. It is the probability which is deemed sufficiently anomalous to update this model and although it is currently twice the cutoff probability to output a result there is nothing that requires this, it was just a reasonable value.

File: lib/maths/CTimeSeriesModel.cc (hunk 2 above)
Comment (tveasey) [reply lib/maths/CTimeSeriesModel.cc]: I'd deleted the restore code from the wrong branch. Good catch.

**Comment by tveasey** [comment]
Ok I agree this is on the fence regarding bug fix vs enhancement. I'm happy to delay until 7.3 and we can test more. I've updated the naming accordingly.

File: docs/CHANGELOG.asciidoc [hunk 3]
```diff
@@ -34,6 +34,7 @@
 
 * Fix an edge case causing spurious anomalies (false positives) if the variance in the count of events
 changed significantly throughout the period of a seasonal quantity. (See {ml-pull}489[#489].)
+* Reduce false positives associated with the multi-bucket feature. (See {ml-pull}491[#491].)
```
Comment (droberts195) [inline docs/CHANGELOG.asciidoc]: Please move this to the 7.3.0 section.

**Comment by Crazybus** [comment]
jenkins test this please
